src/Engine/UnitActionLifecycles/UnitProduceLifecycle.java

```diff
@@ -28,12 +28,32 @@ public UnitProduceFactory(UnitModel type)
       typeToBuild = type;
     }
 
+    // Provide an invalid option if it has materials
+    @Override
+    public GameActionSet getGUIActions(GameMap map, GamePath movePath, Unit actor, boolean ignoreResident)
+    {
+      GameActionSet realActions = getPossibleActions(map, movePath, actor, ignoreResident);
+      if( null == realActions )
+      {
+        XYCoord moveLocation = movePath.getEndCoord();
+        if( moveLocation.equals(actor.x, actor.y)
+            && actor.hasMaterials() )
+        {
+          GameActionSet gas = new GameActionSet(new UnitProduceAction(this, actor), false);
+          gas.isInvalidChoice = true;
+          return gas;
```

Comment:
Remind me what `hasMaterials` is checking - it's `return materials > 0 || !model.needsMaterials;`, but what is `needsMaterials`? It's defaulted to true, which seems wrong if it's a marker for which units can use materials, it isn't referenced anywhere else, and there isn't a comment explaining anything.
Also I tried to see if I could make this invalid option appear, but I'm still (intermittently?) running into the "you cannot build anything" bug.

Reply:
Further inspection implies that `needsMaterials` is a flag for whether the unit should cost a material, if produced by another unit. This is kinda the opposite of what the variable name implies.

Reply:
It's probably worth noting that while reviewing code, you can actually go to the definition by clicking on the function name:
![image](https://github.com/user-attachments/assets/06c60090-2246-46c1-bd20-b783b4be8e14)
> It's defaulted to true, which seems wrong if it's a marker for which units can use materials, it isn't referenced anywhere else

It's a marker for which units can run out of materials, i.e. have finite materials. 0 is a finite number of materials.
I put that in for KaijuWars because the sky carrier needs to be able to produce planes forever (and evidently forgot to make DoR APCs building things use `hasMaterials()`, derp).

Your point of "not referenced anywhere" is kind of a flaw in Lombok - the generated builder/setter for that value are how it gets set, and you can't see the definitions (or find all references unless you've found at least one).

I suppose an equivalent option would be to provide versions of those actions that don't cost materials, but this seemed simpler.
> there isn't a comment explaining anything.

Touche. IIRC we discussed the name at length, but a name isn't always enough.

> Also I tried to see if I could make this invalid option appear, but I'm still (intermittently?) running into the "you cannot build anything" bug.

You need a DoR Carrier, materials, and no money/unit count for a Seaplane.
As for the "bug", I still think that's saved options biting you (because now the order has changed, so any options you saved in the interim would have an unrelated value in the final slot). Try deleting `res/map_options`.
> Further inspection implies that `needsMaterials` is a flag for whether the unit should cost a material, if produced by another unit. This is kinda the opposite of what the variable name implies.

I have no idea how you came up with that.
`needsMaterials` is only used in `hasMaterials()`, which is called on the actor (the unit that exists that wants to build another unit) to find out whether it has materials.
A DoR Carrier normally has 4 materials. If we changed it to have `needsMaterials = false`, then it would have infinite materials.

Reply:
> You need a DoR Carrier, materials, and no money/unit count for a Seaplane.

So after running out of materials entirely, even the invalid menu option disappears. Also, do we have a visual display for materials? I don't see one (or remember adding one). Be a good issue to create if not (as if we ever revisit those).

> As for the "bug", I still think that's saved options biting you

That fixed it. BUT, is this also going to bite every person online who slaps in the new jar? Is there some way we can detect the old version and update (or just ignore) it?

> I have no idea how you came up with that.

On further further inspection, I think I misread `performEvent`, which checks `builder.model.needsMaterials`. I think I saw `model`, and assumed `builder` was the Lombok thing, instead of, y'know, the unit doing the building. Sigh.

> needsMaterials is only used in hasMaterials()

And `UnitProduceLifecycle.performEvent()`. Heh.

Reply:
> So after running out of materials entirely, even the invalid menu option disappears.

Right. The invalid option says "if the state of the world changes a bit, you could be building a seaplane right now".
Greyfield is the only way to restore materials, though.
...on that note, should Jess restore materials? I feel like I should make her restore materials.
> Also, do we have a visual display for materials? I don't see one (or remember adding one). Be a good issue to create if not (as if we ever revisit those).

Nnnnnope. I think I considered it and didn't feel like making an icon.
> That fixed it. BUT, is this also going to bite every person online who slaps in the new jar?

It would have, I think.
> Is there some way we can detect the old version and update (or just ignore) it?

Yes, that's the new feature that gave you such problems just now.
It now orders the *saved* items in the order they were added to the code, instead of menu order.
This just happened to cause problems for people unlucky enough to run the code between adding the new option and changing that order.

> On further further inspection, I think I misread `performEvent`, which checks `builder.model.needsMaterials`. I think I saw `model`, and assumed `builder` was the Lombok thing, instead of, y'know, the unit doing the building. Sigh.

Heh, fair enough.